Clean and improve identity API

lzap opened the pull request:
Hey,

the identity API is currently based only on middleware. This does not work great in regard to logging and observability. We were running into issues when identity was incorrect and all we see in logs was 400. Apparently, logging full body payloads is not viable strategy for production environments and the identity package does not perform any logging, except one line through go standard library "log" package, which is not ideal.

This patch breaks the middleware into two separate exported functions `DecodeIdentity` and `DecodeIdentityCtx` which can be used separately in a small middleware in the individual apps. They properly return identity/context and all possible errors so applications can decide on more detailed logging and monitoring. The EnforceIdentity middleware behavior is left unchanged.

Applications can now implement their own logging or other features they need:

```go
func EnforceIdentity(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx, err := DecodeIdentityCtx(r.Context(), r.Header.Get("X-Rh-Identity"))
		if err != nil {
			// perform own logging, metrics etc
			http.Error(w, http.StatusText(400)+": "+err.Error(), 400)
			return
		}
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}
```

The only behavioral change is that the payload error message changed from:

`Bad Request: x-rh-identity header does not contain valid JSON`

to

`Bad Request: x-rh-identity header does not contain valid JSON: invalid character '}' after top-level value`

It now includes the JSON parser error message. Other than that, there should be no changes to existing applications.

lzap commented on go.mod:
```diff
@@ -3,12 +3,13 @@ module github.com/redhatinsights/platform-go-middlewares
+	go.uber.org/multierr v1.9.0 // indirect
+	go.uber.org/zap v1.24.0
+	golang.org/x/sys v0.4.0 // indirect
 )
```
I had to update dependencies since x/sys was not building on Apple Silicon.

lzap commented on identity/identity.go:
```diff
@@ -103,66 +99,91 @@ func GetIdentityHeader(ctx context.Context) string {
+	if identity.Identity.OrgID == "" && identity.Identity.Internal.OrgID != "" {
+		identity.Identity.OrgID = identity.Identity.Internal.OrgID
+	}
+}
```
This was moved from the bottom.

lzap commented on identity/identity.go:
```diff
@@ -103,66 +99,91 @@ func GetIdentityHeader(ctx context.Context) string {
+	DecodeIdentityError       = errors.New("unable to b64 decode x-rh-identity header")
+	InvalidOrgIdIdentityError = errors.New("x-rh-identity header has an invalid or missing org_id")
+	MissingIdentityTypeError  = errors.New("x-rh-identity header is missing type")
+)
```
The package now export and return these errors. The middleware still writes them into body with HTTP 400 code to keep the old behavior but callers can implement their own middleware with the newly introduced exported functions.

subpop replied:
I don't have any hard evidence to back this up, but the pattern I have observed throughout the Go stdlib is that when naming an error type (i.e. in the type declaration), the pattern `FooError` is used:

```go
type IdentityError struct {
    description string
    wrappedError error
}

func (e *IdentityError) Error() string {
    var s string
    if e.wrappedError != nil {
        s = fmt.Sprintf("%w", e.wrappedError)
    }
    if description != "" {
        s += ": " + description
    }
    return s
}
```

Then, when declaring a variable of that type, the pattern `ErrSpecificDescription` is used:

```go
var ErrMissingIdentityHeader = IdentityError{description: "missing x-rh-identity header"}
```

This would allow users to compare errors using the `errors.Is` and `errors.Wraps` API. I don't think that API can be used directly with `errors.errorString` (the type returned by `errors.New`.

lzap replied:
Up until now, I have been mixing those. This make sense, I like that and I will start enforcing this on my projects too.

lzap commented on identity/identity.go:
```diff
@@ -103,66 +99,91 @@ func GetIdentityHeader(ctx context.Context) string {
+	}
+
+	return xRhId, nil
+}
```
This was extracted from the middleware function, instead of writing to HTTP body it now returns error type.

subpop commented on identity/identity.go:
```diff
@@ -75,20 +75,16 @@ type XRHID struct {
-
 // Get returns the identity struct from the context
 func Get(ctx context.Context) XRHID {
 	return ctx.Value(Key).(XRHID)
```
This panics the program if the typecast fails. Could you include a simple check on the cast and return a zero-value XHRID struct instead?

```suggestion
        x, ok := ctx.Value(Key).(XRHID)
        if !ok {
            return XRHID{}
        }
	return x
```

lzap replied:
Absolutely.

lzap replied:
Although I have one idea - the reason why you can ever see panic is because `Key` is an exported value. It should not have been ever exported, this allows callers to mess around with the value. I am going to change my patch to unexport this key for v2, meaning there is no way to be able to save a value into the context under this key for anyone except this package. Therefore panic will never happen.

The reason for others to use the Key directly was I think because the API was not designed cleanly and there was no good way of implementing your own middleware easily. This will be actually now possible using many new exported functions which will allow others to implement middleware easily.

What you think? Is my understanding of this problem correct?

lzap commented:
Note to myself: take a look on other implementations I was made aware of if there is anything I can pull in here too:

* https://github.com/RedHatInsights/module-update-router/blob/master/identity/identity.go
* https://github.com/RedHatInsights/identity/blob/main/identity.go

lzap commented:
Since we are going for V2 version and bigger changes are possible, here is my plan:

* Deprecate `Get/With` functions, replace with `Identity/WithIdentity` which is in alignment with Go stdlib context naming policy (no "getters").
* Introduce new key and `RawIdentity/WithRawIdentity` that will carry the raw identity (unparsed one) for faster handover in apps.
* Deprecate `GetIdentityHeader`, replaced with `RawIdentity`.
* Unexport `Key`, it should really not be exported. The interface are the functions, not keys.
* Separate parsing and validation into two exported functions so apps can decide if to validate or to what degree they would like to do it.
* Rename `checkHeader` to `EnforceBasePolicy`.
* Add `Entitlements` type-safe version, meaning 1.18+ Go version is required for v2.

lzap commented:
Pushed the whole refactoring, I recommend to read the code and not diff as there are many changes, I am splitting everything into separate files. Everything I planned is implemented, except that I need to prefix with `Get` so `GetIdentity` is the function name because there is already a type `Identity` in the package.

To be done:

* [ ] more unit tests for newly introduced functions.
* [ ] entitlement field (should I add it? I am not sure what this is for)

subpop commented on identity/ctx_keys.go:
```diff
@@ -0,0 +1,9 @@
+type identityKey int
+
+const (
+	parsedKey identityKey = iota
```
What is your reason for putting identityKey into a separate file? Isn't it basically another type?

lzap replied:
No other reason than breaking it apart a bit. Could be probably in the identity.go file as well, yeah. You want me to put it back?

subpop replied:
That would be my preference. I don't really like the pattern of separating types out into a file apart from methods or functions that use those types. I find it a lot easier to reason with if the type is as close to the methods or functions that use that type as possible, and a separate file just creates an arbitrary separation between them.

lzap replied:
Working on it.

subpop reviewed (changes requested):
Could you include a README in this package itself that describes how to use it? The package doc comments in identity.go are a good start, but having a visible README could make adoption of the package easier for people.

lzap commented:
Added a README which links the official documentation site and `package.go` with the full documentation which is the de-facto standard for longer documentation of packages. Let me know if you like this and I will do the unit tests.

lzap commented:
Added more tests, incorporated your remarks, moved identity keys back to the original file (I am assuming breaking other functionality into separate is fine), added v2 prefix to the module name. I think this might be it.

lzap commented:
@subpop I know it is bad time of the year, maybe in a month we can revive this effort? I would love to push this forward.

lzap commented:
Any chance you can get back to it? @subpop

jhjaggars commented:
@coderbydesign are you aware of this PR?

subpop commented on identity/deprecated.go:
```diff
@@ -0,0 +1,54 @@
+package identity
```
What is the purpose of moving these functions to a separate file? It makes the
change harder to trace. If the functions are deprecated, add a deprecation
comment (like you did), but leave the functions in place. That way, it's clear
that the only change to the function is the addition of the deprecation comment.

lzap replied:
Moved back.

subpop commented on identity/package.go:
```diff
@@ -0,0 +1,33 @@
+	r := mux.NewRouter()
+	r.Use(identity.EnforceIdentityWithLogger(ErrorLogFunc))
+*/
+package identity
```
I don't like the idea of a separate "package.go" file for a package this small.
This package comment could easily exist at the top of `identity.go`. It's not
that large of a package. This keeps the documentation closer to the code too,
which is usually a good thing.

lzap replied:
Moved back.

subpop commented on identity/identity.go:
```diff
@@ -103,66 +64,115 @@ func GetIdentityHeader(ctx context.Context) string {
+	var id XRHID
+	err = json.Unmarshal(idRaw, &id)
+	if err != nil {
+		return XRHID{}, fmt.Errorf("x-rh-identity header does not contain valid JSON: %w", err)
```
Could the errors be organized into a type like `IdentityError` and this become a
`IdentityUnmarshalError` error?

lzap replied:
Wrapping multiple errors is only supported from Go 1.22, this library is 1.16+ and I do not want to change that in this PR. I am not a huge fan of go.uber.org/multierr. I think it is a good idea anyway, so let me do it in a way that this will be ready for multi-error in Go 1.22:

```
	// TODO: use multiple error wrapping once Go 1.22 compatibility is reached
	return XRHID{}, fmt.Errorf("%w: %s", ErrUnmarshalIdentity, err.Error())
```

For introducing a new type, I do not see what it brings to the table. I am keeping using the standard library `error` type.

lzap replied:
_I edited the above comment, I was wrong about the Go version which introduced multi-error wrapping. Time flies! :-)_

lzap replied:
Alternatively, I can bump to 1.20 and take advantage of multi-error wrapping in this case. Since 1.20 is already in ubi8/ubi9 and because this is a v2 version bump, I think we could afford such bump.

The only concern about that is identity is not the only library part of this repository, there is also logging and request_id. We don’t use these in our app, both only export a dozen of lines. I can imagine squeezing some incompatible v2 changes into these packages as well:

* unexport https://github.com/RedHatInsights/platform-go-middlewares/blob/master/request_id/request_id.go#L21
* upgrade to aws sdk v2 https://github.com/RedHatInsights/platform-go-middlewares/blob/master/logging/cloudwatch/hook.go

subpop replied:
This seems reasonable to me. This package is intended to be used in server applications. Those applications can readily increment the version of Go they're using, and if 1.20 is available in ubi8 and ubi9, it's a reasonable baseline.

subpop commented on identity/identity.go:
```diff
@@ -103,66 +64,115 @@ func GetIdentityHeader(ctx context.Context) string {
-		rawHeaders := r.Header["X-Rh-Identity"]
+// If org_id is not defined at the top level, use the internal one. See:
+// https://issues.redhat.com/browse/RHCLOUD-17717
+func topLevelOrgIDFallback(identity *XRHID) {
```
This function is only called once, inside `DecodeIdentity`. Is there any
particular reason why it needs to exist outside that function as a separate one?

subpop commented on identity/identity.go:
```diff
@@ -103,66 +64,115 @@ func GetIdentityHeader(ctx context.Context) string {
 
+// CheckBasePolicy performs semantic identity check and returns nil for valid values or
+// an error instead.
+func CheckBasePolicy(id *XRHID) error {
```
This is an exported function. Is the intention that a user will know what to use
it for? It's too is only called once within `DecodeIdentity`; does it need to
exist as a standalone function?

lzap replied:
I extrapolated the approach, I have nothing in my hands but I feel like it might be useful. But if the end result is a cleaner API then I agree, un-exported it.

subpop reviewed (changes requested):
This is looking great to me. I have a few thoughts/questions that came up as I reviewed this again. Overall, I like the pattern of the package how it builds on DecodeIdentity; this makes it clear to the user they they could pick the level of middleware functionality that suites their need.

lzap commented:
Rebased on top of dependencies updates, resolved all the mentioned remarks, added Go 1.19 as the new minimum Go version and bumped all dependencies to the latest versions.

subpop commented on identity/identity.go:
```diff
@@ -103,66 +144,179 @@ func GetIdentityHeader(ctx context.Context) string {
+	if value == nil {
+		return XRHID{}
+	}
+	return value.(XRHID)
```
How confident are you that this won't panic the program? Wouldn't this be safer?

```suggestion
	if v, ok := value.(XRHID); ok {
	    return v
	} else {
	    return XRHID{}
	}
```

lzap replied:
Pretty confident because the context key type named `parsedKey` is now not exported, thus the only code that can insert such value into a context is few lines below: `context.WithValue(ctx, parsedKey, id)` and you cannot pass `nil` into this function because `id` is not a pointer.

subpop replied:
No, I'm not going to insist. You make a reasonable argument for the decision to explicitly unwrap the value as an `XRHID` type. I agree the most likely scenario encountered here is the absence of the header, which you account for by checking whether the value is nil already.

lzap commented:
Rebased, minimum Go version is now 1.20. Took advantage of multi-error wrapping. Unexported the context key in `request_id` package. For the AWS SDK v2 upgrade, I am not going to do it in this PR because it is a bit more work and it is not an incompatible change after all.

I believe this should be safe to merge, it is a v2 package and Go tooling will ignore it for existing apps.

lzap commented:
What I suggest to do:

* tag the HEAD with v1.0.0
* merge this PR
* review other PRs and merge these which are ready
* tag v2.0.0

subpop commented:
Let's do it!

lzap commented:
Awesome, thanks!
